core/src/main/scala/kafka/server/ConfigHandler.scala

```diff
@@ -62,6 +62,12 @@ class TopicConfigHandler(private val logManager: LogManager, kafkaConfig: KafkaC
     topicConfig.asScala.forKeyValue { (key, value) =>
       if (!configNamesToExclude.contains(key)) props.put(key, value)
     }
+
+    if (!kafkaConfig.remoteLogManagerConfig.enableRemoteStorageSystem()
```

Comment:
Correct me if I am wrong here but this method is called when each broker is trying to implement the configuration change sent to it by metadata topic (kraft) or zk. Even if we reject it here, the config change will continue to persist in Zk and metadata.

We need to block this configuration change at the start itself, i.e. in KafkaConfig.scala

Reply:
Ah, I realised that this method is also called during startup and the `kafkaConfig.remoteLogManagerConfig` is not stored in Zk since it is not a dynamic config.

I was hoping we could do something like https://github.com/apache/kafka/blob/8dec3e66163420ee0c2c259eef6e0c0f3185ca17/core/src/main/scala/kafka/server/KafkaConfig.scala#L2212 . This is because we currently don't perform config validation in ConfigHandler and I was hoping there is a better place to do it.

Let me think about it and will let you know if I have a better suggestion.

Reply:
If you are referring to the `emote.log.storage.system.enable` property it is a read-only property and is only stored in server.properties (i.e. not in Zookeeper).

Reply:
In essence, we cannot block the configuration if the property is read-only. What we can do is terminate the startup of the broker should we detect that this property is set to false and there are topics with a `remote.storage.enable` property set (regardless of whether it is true or false).

Reply:
> Ah, I realised that this method is also called during startup and the `kafkaConfig.remoteLogManagerConfig` is not stored in Zk since it is not a dynamic config.
> 
> I was hoping we could do something like
> 
> https://github.com/apache/kafka/blob/8dec3e66163420ee0c2c259eef6e0c0f3185ca17/core/src/main/scala/kafka/server/KafkaConfig.scala#L2212
> . This is because we currently don't perform config validation in ConfigHandler and I was hoping there is a better place to do it.
> 
> Let me think about it and will let you know if I have a better suggestion.

I saw that as well, but the problem is that the topic configuration is a separate entity to the KafkaConfig. But sure, let me know if something else has your preference.

Reply:
For my understanding, will this approach will work if the user wants to downgrade from 3.6 to 2.8?

Reply:
Sorry, can you elaborate because I am not certain I understand?

Reply:
We are ensuring that the system level remote storage cannot be disabled when the topics are enabled with remote storage. If the user downgrades the build, then this validation won't be applicable.

Reply:
@clolov could you please check if we call some form of LogConfig.validate() on startup? (I think we do when we create LogManager). In which case you may want to bring the broker level config to LogConfig like this: https://github.com/apache/kafka/pull/14176 and add you new validation in LogManager.validate() similar to how we are validating in other PR.

Reply:
I actually prefer that we keep values from KafkaConfig (system-wide Tiered Storage configuration) and LogConfig (topic-specific Tiered Storage configuration) logically separate rather than insert one in the other. In other words, I would prefer that the validation between KafkaConfig(s) and LogConfig(s) happens in a place where values can be pulled from both and compared with one another (like what this PR is doing) until we introduce a mechanism of dependence between configurations from different sources.

Of course, if I have missed a KRaft configuration which appears to have a similar logic as what we are trying to achieve here and it has been put in the LogConfig I can be persuaded to change my mind.

My reasoning comes from default values (log.segment.bytes and log.segment.ms) in KafkaConfig being pulled from more specific configs (for example the segment.bytes and segment.ms) rather than the values from KafkaConfig being supplied to the LogConfig and being overwritten there.

Reply:
I think this is the place we can get all custom configurations in topics when startup, so this change makes sense to me.

Reply:
Hello @showuon and thanks for looking over the pull request! Just to confirm, you are okay with the solution proposed here i.e. to not put this validation in the LogConfig?

Reply:
It makes sense to me. But we also want to get consensus from all reviewers.

Reply:
With this change we are splitting the logic of system-wide and topic-specific config at two places. One in LogConfig and this one in ConfigHandler. Could we choose to stick to one of them? I don't have a strong opinion on which one as long as it's at one consolidated place.

Reply:
I have been looking at this a bit more and I don't think the current codebase invokes the validation path on pulling dynamic configurations. In other words, I believe there is client to broker validation, broker to metadata store validation, but once the data makes it in the metadata store we no longer carry out validation when we read it back from there. Unless I am wrong this means that we cannot unite these two approaches without big changes (system-wide needs to happen when we pull the dynamic configuration while the topic-specific needs to happen at broker to metadata store boundary). Am I wrong?